lib/sample_rate_converter/sample_rate_converter.c

```diff
@@ -0,0 +1,357 @@
+	size_t bytes_per_sample = sizeof(uint16_t);
+#elif CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_32
+	size_t bytes_per_sample = sizeof(uint32_t);
+#endif
+
+	size_t samples_in = input_size / bytes_per_sample;
+
+	if ((ctx->input_sample_rate != input_sample_rate) ||
+	    (ctx->output_sample_rate != output_sample_rate) || (ctx->filter_type != filter)) {
+		LOG_DBG("State has changed, re-initializing filter");
+		ret = sample_rate_converter_init(ctx, input_sample_rate, output_sample_rate,
+						 filter);
+		if (ret) {
+			LOG_ERR("Failed to initialize converter (%d)", ret);
+			return ret;
+		}
+	}
+
+	conversion_ratio = calculate_conversion_ratio(input_sample_rate, output_sample_rate);
+
+	if (!conversion_needs_buffering(ctx, conversion_ratio)) {
+		write_ptr = output;
+		read_ptr = input;
+		samples_to_process = samples_in;
+	} else {
+		read_ptr = internal_input_buf;
+		write_ptr = internal_output_buf;
+
+		if (((samples_in + (ctx->input_buf.bytes_in_buf * bytes_per_sample)) %
+		     conversion_ratio) == 0) {
+			size_t extra_samples = conversion_ratio - (samples_in % conversion_ratio);
+
+			LOG_DBG("Using %d extra samples from input buffer", extra_samples);
+			samples_to_process =
+				samples_in + (conversion_ratio - (samples_in % conversion_ratio));
+		} else {
+			size_t extra_samples = (samples_in % conversion_ratio);
+
+			LOG_DBG("Storing %d samples in input buffer for next iteration",
+				extra_samples);
+			samples_to_process = samples_in - extra_samples;
+		}
+
+		/* Merge bytes in input buffer and incoming bytes into the internal buffer for
+		 * processing
+		 */
+		memcpy(internal_input_buf, ctx->input_buf.buf, ctx->input_buf.bytes_in_buf);
+		memcpy(internal_input_buf + ctx->input_buf.bytes_in_buf, input, input_size);
+	}
+
+#ifdef CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_16
+	if (is_upsampling(ctx->input_sample_rate, ctx->output_sample_rate)) {
+		arm_fir_interpolate_q15(&ctx->fir_interpolate_q15, (q15_t *)read_ptr,
+					(q15_t *)write_ptr, samples_to_process);
+		bytes_produced = samples_to_process * conversion_ratio * bytes_per_sample;
+	} else {
+		arm_fir_decimate_q15(&ctx->fir_decimate_q15, (q15_t *)read_ptr, (q15_t *)write_ptr,
+				     samples_to_process);
+		bytes_produced = (samples_to_process / conversion_ratio) * bytes_per_sample;
+	}
+#elif CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_32
+	if (is_upsampling(ctx->input_sample_rate, ctx->output_sample_rate)) {
+		arm_fir_interpolate_q31(&ctx->fir_interpolate_q31, (q31_t *)read_ptr,
+					(q31_t *)write_ptr, samples_to_process);
+		bytes_produced = samples_to_process * conversion_ratio * bytes_per_sample;
+	} else {
+		arm_fir_decimate_q31(&ctx->fir_decimate_q31, (q31_t *)read_ptr, (q31_t *)write_ptr,
+				     samples_to_process);
+		bytes_produced = (samples_to_process / conversion_ratio) * bytes_per_sample;
+	}
+#endif
+
+	if (!conversion_needs_buffering(ctx, conversion_ratio)) {
+		/* Nothing needs to be done in output buffer */
+		return 0;
+	}
+
```

Comment:
Discussed offline, and we agree on the ringbuffer method gives a lot of unnecessary shuffling, but it was chosen as a more specialized buffer solution was more complex to implement than first thought, and the fact that the ring buffer is a well tested existing solution that works. Making a specialized solution is definitely a future improvement though, as it would reduce the context size by the block size (~1k).